### ISSUE TITLE
"strict" interface implementation.

### ISSUE THREAD

**Issue by MkazemAkhgary** [dotnet/csharplang#914]
**Summary**

Interfaces can be re-implemented by derived classes, this is also documented here.

https://msdn.microsoft.com/en-us/library/aa664594(v=vs.71).aspx

This sounds good, but not sometimes, especially when you are designing an API, sometimes you don't want to let others to change interface implementations, and that's where problems arises. you may think to make interface internal, or make classes sealed or even changing the design of your API entirely. 

**Basic example**

```
interface IControl
{
   void Paint();
}
class Control: IControl
{
   void IControl.Paint() {...} // control is extendable, but please. not this one!
}
class MyControl: Control, IControl
{
   public void Paint() {}
}
```

**Proposal**

we can fix this issue by using a keyword to denote that this interface can not be re implemented in derived classes.

```
interface IControl
{
   void Paint();
}
class Control: strict:IControl
{
   void IControl.Paint() {...} // control is extendable, but please. not this one!
}
class MyControl: Control, IControl // You shall not pass!
{
   public void Paint() {}
}
```

Multiple strict implementations along with normal implementations.

`class Control: IControl1 strict : IControl2, IControl3`

strict must come before `where` constraint and after inheriting base class. any interface after strict: keyword is implemented strictly. 

if interface is named strict it will not cause conflicts. 

`class MyClass : strict` in order to become strict must be `class MyClass : strict:strict`

Another point of interest. here interface call can be optimized too like how sealed method calls are optimized.

**Comment by HaloFour** [comment]
There's no existing facility in the CLR which would enable the C# language to enforce such a rule.

**Comment by yaakov-h** [comment]
@HaloFour There's also no existing facility in the CLR for DIM or other language suggestions...

**Comment by Joe4evr** [comment]
But most other suggestions can compile down to an existing CLR construct which can make it enforceable at runtime. This proposal is something that requires runtime support to work, but beyond that, it breaks polymorphism because a derived class should be able to provide a more specific implementation of something.

**Comment by MkazemAkhgary** [comment]
@Joe4evr why it breaks polymorphism?, if so then why sealed methods would be a thing? I was thinking of making interface implementations sealed. anyhow good too know this is not independent from runtime. I thought this is something related only to compile time.

**Comment by svick** [comment]
@MkazemAkhgary 

> anyhow good too know this is not independent from runtime. I thought this is something related only to compile time.

I think there are two options:

1. Change the runtime so that it understands `strict` and ensures it's never broken.
2. Change only the compiler. This would mean that other compilers (including older versions of the C# compiler) would let you reimplement `strict`ly implemented interfaces.

Though if option 2 is acceptable for you, an easier variant would be to create an analyzer (and somehow mark strictly implemented interfaces using an attribute). That way, you would get a similar experience than fully implemented option 2, but without having to change the language.

**Comment by jnm2** [comment]
I'm not really sure I like this. How is it indispensable enough to gain back the -100 points?

**Comment by MkazemAkhgary** [comment]
you wouldn't, until you feel the need for it :) @jnm2

**Comment by jnm2** [comment]
I just don't understand how this comment ever makes sense: `control is extendable, but please. not this one!` More context, please?

**Comment by MkazemAkhgary** [comment]
Ok, have you tried to write an API for specific application development? if you do you probably know sometimes you don't want to expose implementation details. but some times you cant avoid that. in my case when using WinRt where some interfaces are public. I don't want let others to change implementation details of some methods. but since I have no control over that interface to make it internal, I'm left with commenting please don't do this. other wise you may mess things up. @jnm2 

the purpose for this post is not just for adding a feature in language. its not just coding, this is useful when designing application programing interface.

**Comment by jnm2** [comment]
@MkazemAkhgary Or you can do what Castle Windsor does and put the internal implementation methods into an `IKernalInternal` interface- something with `Internal` in the namespace or name.

But have you ever used an API for specific application development and wanted to extend one tiny thing but the API designers decided not to let you override one specific method that would make everything so simple and beautiful, leaving you with either ugly workarounds or a complete block?

**Comment by jnm2** [comment]
The interface methods don't show up by accident.

**Comment by MkazemAkhgary** [comment]
That's a challenge when designing a good API. but remember, when you open a path to new approaches for developers, you can not change that back other wise it would be a breaking change that makes every one complain. so if you make mistake you can not revert. unless you don't really care about who or how many people are using your libraries.

but if you do it carefully, that pretty much wont happen. you can get feed backs, do unit tests and carefully analyze different scenarios so later you can decide how to open path for new approaches. then you have nice and safe API. 

This is pretty much opinion based. yes I agree with your statement but again I prefer safer approach. :-)

thanks for suggestion. ill try that too. @jnm2

**Comment by HaloFour** [comment]
@yaakov-h

I know.  My point is that such a proposal requires an accompanying CLR modification.  That does increase the effort necessary for implementation, possibly beyond the point where it's reasonable, but I'm not suggesting that automatically disqualifies the proposal.

The fact that CLR mods are required for DIM is already a known quantity.

**Comment by yaakov-h** [comment]
I’m with @jnm2 on this. I can’t actually think of a single place where i would want to use this feature.

Can you please provide a concrete use-case?

And what would you do about wrapper classes that could, for example, independently implement the interface, and pass through on some or all methods? Does this come into play at all?

**Comment by MkazemAkhgary** [comment]
Being able to re implement interface can cause problems. I'm not sure why this is not clear?

a very simple example. (after that ill show another problem close to my issue)

```
    public interface IA
    {
        void Write();
    }

    public class A : IA
    {
        void IA.Write()
        {
            Console.WriteLine("A");
        }
    }

    public class B : A, IA
    {
        public new void Write()
        {
            Console.WriteLine("B");
        }
    }
```

Now what this outputs?
```
var b = new B();
b.Write();
((IA)b).Write();
```
It will output
```
B
B
```

This is not acceptable for me. Behavior is defined. how ever you can completely change purpose of base class. you can change Apples to Oranges. is this _always_ ok? No. 

```
public interface WinRt_Interface // that you don't have control over.
{
    IBuffer Data { get; }
}
```

Now imagine you use this buffer to write data into disk. if you change interface implementation, then you end of something else. buffer is not what you expect to be. it can be anything else. 

Problem goes worse. when I decided to follow flyweight pattern and do some caching. caching data is sensitive and now you can have really anything as buffer. caching certainly wont work there.

Basically you have no control over what developer can do with this interface. which part is not clear? yes its good to be able to change interface implementation. but not always. 

Anyway I'm either forced to change my design or live with it. I hope you got my point. @yaakov-h 

Also another point of concern. if you don't implement interface explicitly while behavior is defined but you might not get what you expect. what would happen? possibly a bug.

This is not my case, because I implement that interface explicitly, but just adding more context.

```
var b = new B();
b.Write();
((A)b).Write(); // if interface were not implemented explicitly.
((IA)b).Write();
```

This will output

```
B
A // behavior is defined but not always obvious. is it?
B
```

**Comment by eyalsk** [comment]
@MkazemAkhgary 

> sometimes you don't want to let others to change interface implementations, and that's where problems arises. you may think to make interface internal, or make classes sealed or even changing the design of your API entirely.

Is there a reason you can't use an abstract class? alternatively, maybe you should use composition over hierarchies?

> This is not acceptable for me. Behavior is defined. how ever you can completely change purpose of base class. you can change Apples to Oranges. is this always ok?

I think you're missing the purpose of interfaces then because the whole point is to provide a contract where the implementation is not part of the contract, consumers shouldn't care how implementers of an interface implements the behaviour this is an implementation detail; if you want a more strict contract then you can do what I wrote above and use an abstract class.

> Now imagine you use this buffer to write data into disk. if you change interface implementation, then you end of something else. buffer is not what you expect to be. it can be anything else.

That's the whole point, you expect to have a buffer so the consumer shouldn't care where it's stored as long as it gets what it asked for.

> Basically you have no control over what developer can do with this interface. which part is not clear? yes its good to be able to change interface implementation. but not always.

Why would you care what consumers do with interfaces? an interface supposed to be used as a protocol, much like the HTTP protocol defines the communication between the server and the client, does it define who is the server? does it define who is the client? not really.

**Comment by DavidArno** [comment]
Even if this feature were implemented, it would be simple to work around it:
```cs
interface IControl
{
   void Paint();
   void Decorate();
}
class Control: strict:IControl
{
   void IControl.Paint() {...} // control is extendable, but please. not this one!
   void IControl.Decorate() {...} // control is extendable, but please. not this one!
}
class MyControl: IControl 
{
    private readonly Control _control;

    void IControl.Paint() => _control.Paint(); // using the original here
    void IControl.Decorate() {...} // using my own implementation here
}
```

Since it is of questionable worth, and easily circumvented, this doesn't seem like a good proposal to me.

**Comment by MkazemAkhgary** [comment]
@DavidArno I'm writing API over API provided with more abstraction and usage. yes if developer would use WinRT directly, then basically he is not using my API. 

@eyalsk What abstract class? what I'm saying is that I don't have control over WinRt. I can not change those interfaces etc, I have to use interface. but thanks for other explanations, they sound reasonable.

Alright this seems to be not reasonable for some of you, I was trying to shoot my final arrow here. but anyway I can deal with this by finding some workarounds or really get used to it.

**Comment by eyalsk** [comment]
@MkazemAkhgary 

> What abstract class? what I'm saying is that I don't have control over WinRt.

This is a simplistic example that isn't based on any requirements or understanding of what your library/application does so please take it with a grain of salt.

```
void Main()
{
	Control control = new Control();
	
	control.Bar();
}

public abstract class Component
{
	private readonly WinRt_Interface _winRt;
	
	protected Component() : this(/* <default behaviour?> */)
	{
	}
	
	protected Component(WinRt_Interface winRt)
	{
		_winRt = winRt;
	}

	protected virtual void Foo()
	{
		// Do something with _winRt.Data
	}
}

public class Control : Component
{
	public void Bar()
	{
		// Do something with Foo();
	}
}
```

**Comment by MkazemAkhgary** [comment]
I get it now. so all I have to do is to use interface internally. the problem was that I was implementing it.

Even if developer wants to access some WinRT methods directly I could add extension methods which will delegate internal interface into WinRt.

@eyalsk Thank you. I see that this proposal does not worth required effort, so I'm closing it.

**Comment by jnm2** [comment]
@eyalsk Thanks for putting the effort into this one. Implementing the interface on a private nested class rather than implementing it on the main class is a valid (and preferable) technique.

**Comment by eyalsk** [comment]
You welcome guys, my pleasure. 😉
